walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,69 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
+
 
 public final class FindMeetingQuery {
+
+  private interface Criterion {
+    boolean satisfies(int x);
+  }
+
+  private static final int DAY = 60 * 24;
+
+  private int getStart(Event event) {
+    return event.getWhen().start();
+  }
+
+  private int getEnd(Event event) {
+    return event.getWhen().end();
+  }
+
+  private boolean sameAttendees(Collection<String> a, Collection<String> b) {
+    return new HashSet<>(a).removeAll(b);
+  }
+
+  private List<TimeRange> makeSchedule(int [] occupied, Criterion criterion, int duration) {
+    List<TimeRange> schedule = new ArrayList<>();
+    int index = 0;
+    while (index < DAY) {
+      int start = index;
+      while (index < DAY && criterion.satisfies(occupied[index])) {
+        index++;
+      }
+      if (index - start >= duration) {
+        schedule.add(new TimeRange(start, index - start));
+      }
+      while (index < DAY && !criterion.satisfies(occupied[index])) {
+        index++;
+      }
+    }
+    return schedule;
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    int duration = (int) request.getDuration();
+    int [] occupied = new int[DAY + 1];
+
+    for (Event event : events) {
+      if (sameAttendees(event.getAttendees(), mandatoryAttendees)) {
+        for (int i = getStart(event); i < getEnd(event); i++) {
+          occupied[i] = 2;
+        }
+      }
+      if (sameAttendees(event.getAttendees(), optionalAttendees)) {
+        for (int i = getStart(event); i < getEnd(event); i++) {
+          occupied[i] = Math.max(1, occupied[i]);
+        }
+      }
+    }
+
+    List<TimeRange> commonSchedule = makeSchedule(occupied, x -> x == 0, duration);
+    List<TimeRange> mandatorySchedule = makeSchedule(occupied, x -> x <= 1, duration);
+    return commonSchedule.isEmpty() ? mandatorySchedule : commonSchedule;
```

Comment:
tip: if `commonSchedule` is non-empty, you can avoid computing `mandatorySchedule` altogether, saving on a bit of work sometimes.